odoo/addons/base/models/res_users.py

```diff
@@ -1001,30 +1001,55 @@ def action_revoke_all_devices(self):
     @api.model
     @api.readonly
     def has_group(self, group_ext_id):
-        # use singleton's id if called on a non-empty recordset, otherwise
-        # context uid
-        uid = self.id
-        if uid and uid != self._uid:
-            self = self.with_user(uid)
-        return self._has_group(group_ext_id)
+        """Return True if the user contains the required group."""
+        # TODO: Add to log a warning when the group_id is None. Because, this implies that the
+        # group xml id does not exist and therefore the data may be inconsistent.
+        # This change must be made at the same time as the __manifest__ and xml changes because
+        # several modules use groups before those are created.
+        group_id = self.env['ir.model.data']._xmlid_to_res_id(group_ext_id, raise_if_not_found=False)
 
-    @api.model
-    @tools.ormcache('self._uid', 'group_ext_id')
-    def _has_group(self, group_ext_id):
-        """Checks whether user belongs to given group.
-
-        :param str group_ext_id: external ID (XML ID) of the group.
-           Must be provided in fully-qualified form (``module.ext_id``), as there
-           is no implicit module to use..
-        :return: True if the current user is a member of the group with the
-           given external ID (XML ID), else False.
-        """
-        assert group_ext_id and '.' in group_ext_id, "External ID '%s' must be fully qualified" % group_ext_id
-        module, ext_id = group_ext_id.split('.')
-        self._cr.execute("""SELECT 1 FROM res_groups_users_rel WHERE uid=%s AND gid IN
-                            (SELECT res_id FROM ir_model_data WHERE module=%s AND name=%s AND model='res.groups')""",
-                         (self._uid, module, ext_id))
-        return bool(self._cr.fetchone())
+        # use singleton's id if called on a non-empty recordset, otherwise context uid
+        return group_id in (self or self.env.user)._user_groups()
+
+    def has_group_set(self, group):
+        if not self:
+            return False
+
+        is_str = isinstance(group, str)
+        if is_str and',' not in group:
+            return self.has_group(group)
+
+        group_theory = self.env['res.groups']._get_group_theory()
+        group = group_theory.define(group) if is_str else group
+        if group.is_empty():
+            return False
+        if group.is_univers():
+            return True
+
+        return group_theory.define_from_ids(self._user_groups(), keep_smallers=True) in group
```

Comment:
Before it was the user in the group, but I didn't take it out of the set object because it made a reference to the user object which didn't make much sense.
I took the meaning as the elements of the set groupA can be contained in the set groupB.
Technically the method does the same thing as a `not (groupA & groupB).is_empty()`. (with optimization for simple cases of union with a single leaf)